Implement bhBarcodeScanner Component

jniles opened the pull request:
This PR implements the `bhBarcodeScanner` component and uses it to replace the barcode scanning functionality on the Cash Payments page.

**Component Signature**
```html
<bh-barcode-scanner on-scan-callback="callback(record)"></bh-barcode-scanner>
```

**Description**
The `bhBarcodeScanner` uses the underlying `BarcodeService` to query the barcode API and determine which record to return to the client.  If the API returns a success HTTP code, it will call the `onScanCallback` function.

The key difference between this component and the previous implementation is the ability to re-focus the hidden barcode input if focus is lost.  This is accomplished by clicking a button that is only visible once focus is lost from the hidden input.

**In Action**
![2018-03-28_14-44-36](https://user-images.githubusercontent.com/896472/38032906-a0942c82-3296-11e8-9623-dd777c3dd908.gif)

jniles committed:
feat(barcode): implement barcode component

This commit implements a barcode component to make working with barcodes
a bit easier.  It replaces the cash barcode modal's custom logic with
generic logic to be tested.

Key features:
 1. If the input loses focus, the user is notified and a button appears
 to re-establish focus on the hidden input.
 2. Simple API - only a single callback is provided for the success
 case.
 3. Record agnostic.

Closes #1529.

jniles committed:
chore: remove bhFocusOn

This directive has been superseded now that the barcode component is
implemented.

jniles committed:
test(bhBarcodeScanner): unit test component

This commit adds unit tests to the bhBarcodeComponent.

mbayopanda commented on client/src/i18n/en/barcode.json:
```diff
@@ -1,5 +1,11 @@
+    "AWAITING_HTTP" : "Barcode Read!  Looking up Record...",
+    "READ_SUCCESS" : "Found Record",
+    "READ_ERROR" : "Unreadable Barcode!  Please enter the barcode manually.",
+    "LOST_FOCUS" : "The barcode input has lost focus.  Please click \"Read Barcode\" to ready the barcode component.",
```
👍

mbayopanda commented on client/src/modules/cash/cash-form.service.js:
```diff
@@ -121,7 +121,7 @@ function CashFormService(AppCache, Session, Patients, Exchange) {
 
     if (config.patient) {
-      this.setPatient(config.invoices);
+      this.setPatient(config.patient);
```
👍 good catch

mbayopanda commented on test/client-unit/components/bhBarcoderScanner.spec.js:
```diff
@@ -0,0 +1,139 @@
+    $scope.$digest();
+  }));
+
+  // TODO - how do you find a focused element?
```
The TODO message is about to find a hidden input box or a focused element ? Because the test below is about a hidden input.

jniles replied:
Ah, I can see how that isn't clear.

I originally tried to write a test to find a "hidden, focused" input, but never got the focus to work.  The CSS selector `$(':focus')` is supposed to work, but I never got that working...

So it's currently a TODO.  I should probably just make a skipped test for it.

mbayopanda commented on test/client-unit/components/bhBarcoderScanner.spec.js:
```diff
@@ -0,0 +1,139 @@
+    angular.element(input).triggerHandler('blur');
+    $scope.$digest();
+
+    expect(isHidden(btn)).to.equal(false);
```
👍

mbayopanda commented on client/src/i18n/fr/barcode.json:
```diff
@@ -1,5 +1,11 @@
+    "SCAN" : "Scanner le Code à Barres",
+    "AWAITING_INPUT" : "En Attente d'Entrée",
+    "AWAITING_HTTP" : "Code à barres lu! Recherche l'enregistrement....",
+    "READ_SUCCESS" : "Trouvée",
```
Trouvé et non Trouvée (parce que le mot enregistrement est au masculin).

A good message can be : `Enregistrement trouvé` instead of simply `Trouvé`

mbayopanda commented on client/src/i18n/fr/barcode.json:
```diff
@@ -1,5 +1,11 @@
+    "AWAITING_INPUT" : "En Attente d'Entrée",
+    "AWAITING_HTTP" : "Code à barres lu! Recherche l'enregistrement....",
+    "READ_SUCCESS" : "Trouvée",
+    "READ_ERROR" : "Code à barres illisible! Entrez le code à barres manuellement.",
```
for the "READ_ERROR" a good key and message can be :
```
"READ_FAILED" : "Enregistrement non trouvé"
```
and for any errors during the reading : 
```
"READ_ERROR" : "Une erreur est survenue lors de la lecture du code barre"
```

jniles replied:
That is an excellent idea.  I'll expand the API for it.

mbayopanda commented on client/src/js/components/bhBarcodeScanner.js:
```diff
@@ -0,0 +1,69 @@
+        $ctrl.onScanCallback({ record });
+      })
+      .catch(() => {
+        $ctrl.currentStep = steps.READ_ERROR;
```
If the record is not found this code falls in the .catch section, and set the step to an error.

I think if the record is not found the step must be READ_FAILED.

On the server side if we use `db.exec`, on the client we can have : 
```js
if (records.length === 0) {
  $ctrl.currentStep = steps.READ_FAILED;
}
```

and the READ_ERROR must be set only if we have a real error.

mbayopanda commented on client/src/modules/templates/bhBarcodeScanner.html:
```diff
@@ -0,0 +1,37 @@
+
+    <span class="text-danger" ng-if="$ctrl.currentStep === $ctrl.steps.READ_ERROR">
+      <i class="fa fa-danger-sign"></i> <span translate>BARCODE.READ_ERROR</span>
+    </span>
```
Add also a  case for a READ_FAILED

mbayopanda reviewed (commented):
The code is excellent and the test also, i just made some comments about messages according the step.

jniles committed:
feat(barcode): differentiate between 404/400

This commit differentiates between a "NOT FOUND" barcode (doesn't exist
in the database) and all other errors.  In theory, we shouldn't
encounter a "NOT FOUND" error if users are only scanning barcodes that
are generated by the system, unless someone has deleted it.

This also refactors the tests to remove repetition.

mbayopanda commented on client/src/i18n/en/barcode.json:
```diff
@@ -1,5 +1,12 @@
+    "AWAITING_HTTP" : "Barcode Read!  Looking up Record...",
+    "READ_SUCCESS" : "Found Record",
+    "READ_ERROR" : "An error occurred during lookup.  Please use another method to find the record.",
+    "NOT_FOUND" : "Cannot find a record with that code.  Please search for the document manually.",
```
👍 good

mbayopanda commented on client/src/js/components/bhBarcodeScanner.js:
```diff
@@ -0,0 +1,71 @@
+    READ_SUCCESS : 8,
+    READ_ERROR : 16,
+    LOST_FOCUS : 32,
+    NOT_FOUND : 64,
```
💃

mbayopanda commented on client/src/js/components/bhBarcodeScanner.js:
```diff
@@ -0,0 +1,71 @@
+      })
+      .catch(err => {
+        const isNotFound = (err.status === 404);
+        $ctrl.currentStep = isNotFound ? steps.NOT_FOUND : steps.READ_ERROR;
```
👍

mbayopanda commented on server/lib/barcode.js:
```diff
@@ -36,7 +36,7 @@ function generate(receiptIdentifier, uuid) {
 // - returns the full UUID of the entity
 function reverseLookup(barcodeKey) {
-  const code = barcodeKey.substr(0, 2);
+  const code = barcodeKey.substr(0, 2).toUpperCase();
```
great 👍

mbayopanda reviewed (approved):
LGTM
